### PR TITLE
Assignment done

### PR DESCRIPTION
**Pull request by NikuX**

### CODE REVIEW

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# making_forecasts - "Nicolae Dubanco"

# Part 1 - Loading Data
***
**What went well:**

- Good job with importing the provided tools and implementing a good (maybe complex) way of taking in the metrics data from Wikipedia.:

```python

project = 'en.wikipedia'
access = 'all-access'
agent = 'user'
article = 'Veganism'
granularity = 'daily'
start = '20150701'
end = '20190527'

endpoint = f'/metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}'
url = 'https://wikimedia.org/api/rest_v1' + endpoint
response = requests.get(url)
assert response.status_code == 200
df = pd.DataFrame(response.json()['items'])
df = df.rename(columns={'timestamp':'ds', 'views':'y'})
df['ds'] = pd.to_datetime(df['ds'], format='%Y%m%d%H')6570


```

- Good job reading in the weather and setting a variable to parse the dates:

```python

df['Date'] = pd.to_datetime(df['Date'])
df = df.set_index('Date')

```

**Where you could of improved on:**

-Cleaner way of importing data shown below:

```python 
  
project = 'en.wikipedia'
access = 'all-access'
agent = 'user'
article = 'Veganism'
granularity = 'daily'
start = '20150701'
end = '20190527'

endpoint = f'/metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}'
url = 'https://wikimedia.org/api/rest_v1' + endpoint
response = requests.get(url)
assert response.status_code == 200
df = pd.DataFrame(response.json()['items'])
df = df.rename(columns={'timestamp':'ds', 'views':'y'})
df['ds'] = pd.to_datetime(df['ds'], format='%Y%m%d%H')

```

# Part 2 - Baselines
***
**What went well:**

- Graphs are beutiful and well done .:

```python
#Nice Indexing
df = df.reindex(pd.date_range(start='2017-03-06', end='2020-05-23'))
df.plot(figsize=(16,9));

df['Mean Baseline'] = df['Alternative data (finance)'].mean()
df.plot(figsize=(16,9));

df['Naive Baseline'] = df['Alternative data (finance)'].shift(1).ffill()
df.plot(figsize=(16,9));

from fbprophet import Prophet
m = Prophet()
m.fit(df)

fig1 = m.plot(forecast)


```


# What to work on
- [x] Comment more in your code!
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# evaluating_forecasts - "Nicolae Dubanco"

# Part 1 - Loading Data/Analysis 
***
**What went well:**

- Good job with importing in clean, reusable code: 

```python

%matplotlib inline
import pandas as pd
import requests

project = 'en.wikipedia'
access = 'all-access'
agent = 'user'
article = 'Veganism'
granularity = 'daily'
start = '20150701'
end = '20190527'

endpoint = f'/metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}'
url = 'https://wikimedia.org/api/rest_v1' + endpoint
response = requests.get(url)
assert response.status_code == 200
df = pd.DataFrame(response.json()['items'])
df = df.rename(columns={'timestamp':'ds', 'views':'y'})
df['ds'] = pd.to_datetime(df['ds'], format='%Y%m%d%H')



```

- Nice use of datetime and I like the variable you set for the dataframe: 

```python

from fbprophet import Prophet
from fbprophet.plot import add_changepoints_to_plot

holidays = pd.DataFrame({
  'holiday': 'new_years',
  'ds': pd.to_datetime(['2015-01-01', '2016-01-01', '2017-01-01', '2018-01-01', 
                        '2019-01-01, 2020-01-01', '2021-01-01'], utc=True),
  'lower_window': -5,
  'upper_window': 5,
})

model = Prophet(daily_seasonality=False, holidays=holidays)
model.fit(df)
future = model.make_future_dataframe(periods=365*2)
forecast = model.predict(future)
fig1 = model.plot(forecast)
add_changepoints_to_plot(fig1.gca(), model, forecast)
fig2 = model.plot_components(forecast)
```

**Where you could of improved on:**

- Get into a habbit of setting a X and Y title (looks good): 


```python 

ax.set_xlabel("x label")

ax.set_ylabel("y label")


#Your code
train = weather['2010':'2017']
test  = weather['2018':'2019']

#Right way 

train = weather['2015-07-01':'2018-12-31']
test = weather['2019-01-01':'2019-05-10']

```

# Part 2 - Graphing 
***
**What went well:**

- Good job adjusting the changepoint_prior_scale parameter:

```python

import matplotlib.pyplot as plt

scales = [0.01, 0.05, 0.15, 0.50]
for scale in scales:
    model = Prophet(daily_seasonality=False, holidays=holidays, 
                    changepoint_prior_scale=scale)
    model.fit(df)
    future = model.make_future_dataframe(periods=365*2)
    forecast = model.predict(future)
    fig1 = model.plot(forecast)
    add_changepoints_to_plot(fig1.gca(), model, forecast)
    plt.title(f'changepoint_prior_scale={scale}')
    plt.show()

```

- Good showing of daily high through matplotlib(beautiful simplistic code):

```python
# TODO
from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(df_cv, metric='mae')
plt.title('Wikipedia page views for Alternative data finance')

```
- Check out your classmate's code and compare! 

```python


https://github.com/JefNtungila/DS-Unit-2-Regression-1/blob/master/Jef_Ntungila_assignment_evaluating_forecasts.ipynb


```

# What to work on
- [x] Comment more in code 
- [x] Add X and Y labels 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Nicolae Dubanco- "Regression 1 Sprint Challenge"

# Part 1. Predict avocado sales
***
**What went well:**

-Good job imporing data and dataset and checking out shape:

```python
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)

```

## 1.2. Plot the forecast

- Took it a step further with Prophet so I like this code (added simpler line for future case below):
```python


avocado['Date'] = pd.to_datetime(avocado['Date'])
avocado = avocado.set_index('Date')
avocado.plot(figsize=(16,8));

```

- Using Prophet: 

```python


from fbprophet import Prophet
avocado.columns = ['ds', 'y']
m = Prophet()
m.fit(avocado)
future = m.make_future_dataframe(periods=365*2)
forecast = m.predict(future)

#one liner
fig1 = m.plot(forecast)


```




```

##1.5. Plot the cross-validation Mean Absolute Percentage Error: metric='mape' 
***
**What went well:**

- Clean and elegant :

```python


fig1 = m.plot(forecast)

```


** What could be improved: 

```python 
from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(df_cv, metric='mape')

#This is what we use 
```

#Stretch Goal (PARTIAL)

# Part 2 

##2.1. Begin with baselines for regression. Calculate the mean absolute error and 𝑅2 score for a mean baseline.

- Went went the base .head: 


``` python


rent.head()





```

**What could be better

-use this syntax for better insight:
```python

rent.describe()

```


## 2.2. Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
-Good job choosing features and getting to it (I added another way as well)):
```python

#Your Code
X = rent[['bedrooms','bathrooms']]
y = rent['price']


features = ['bedrooms','bathrooms']

target = 'price'

X = rent[features]
y = rent[target]

model.fit(X, y)

y_pred = model.predict(X)



train = rent.query('month <= 5')
test  = rent.query('month == 6')
X_train = train[features] 
y_train = train[target]
X_test  = test[features]
y_test  = test[target]


model.fit(X_train, y_train)


y_pred = model.predict(X_test)



#My Syntax

features = ['bedrooms']
target = 'price'
x = rent[features]
y = rent[target]


## Other option

train = rent.query('month < 6')
test = rent.query('month == 6')
train.shape, test.shape

features = ['bedrooms', 'bathrooms']
target = 'price'

```

##2.3. Choose any two features.

-use this syntax for better insight:
```python

```

2.4. Use scikit-learn to fit a Linear Regression model on the train data.

-Not bad for just starting out to use to scikit. Syntax is not bad, but features could be better ( see below ):
```python

from sklearn.metrics import mean_absolute_error, r2_score
import matplotlib.pyplot as plt

def plot_preds(x, y, y_pred):
    plt.scatter(x, y, label='y_true')
    plt.plot(x, y_pred, label='y_pred')
    plt.legend()
    
    
    mae = mean_absolute_error(y, y_pred)
    r2 = r2_score(y, y_pred)
    print('Mean Absolute Error:', mae)
    print('R^2:', r2)
    
    
    
plot_preds(x, y, y_pred)
plt.title('Mean Baseline');

```

**What could be better

-better syntax:


```python

# Choose a class of model by importing the appropriate estimator class from Scikit-Learn.
from sklearn.linear_model import LinearRegression

# Choose model hyperparameters by instantiating this class with desired values.
model = LinearRegression()

# Arrange data into a features matrix and target vector.
X_train = train[features]
y_train = train[target]
X_test = test[features]
y_test = test[target]

X_train.shape, y_train.shape, X_test.shape, y_test.shape

# Fit the model to your data by calling the fit() method of the model instance.
model.fit(X_train, y_train)

# Apply the Model to new data: For supervised learning, often we predict labels for unknown data using the predict() method.
y_pred = model.predict(X_test)

```

2.6. Get regression metrics MAE and 𝑅2 for the test data.

-Great job:
```python


mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print('Test Mean Absolute Error:', mae)
print('Test R^2 Score:', r2)

```

2.7. Get the model's coefficients and intercept.

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

-use this syntax for better insight:
```python
model.intercept_, model.coef_
```

#Stretch Goals_2

```python

import seaborn as sns

target = 'price'
features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
for feature in features:
    sns.lmplot(x=feature, y=target, data=train, scatter_kws={'alpha':0.01})
    
    ```
    
    
    
    ```python
    
    
    def fit(train, test, features):
    X_train = train[features]
    y_train = train[target]
    X_test = test[features]
    y_test = test[target]

    model = LinearRegression()
    model.fit(X_train, y_train)

    print(f'Linear Regression with {len(features)} features: {features}')
    print('Train Mean Absolute Error:', 
          mean_absolute_error(y_train, model.predict(X_train)))
    print('Train R^2 Score:', 
          r2_score(y_train, model.predict(X_train)))
    print('Test Mean Absolute Error:', 
          mean_absolute_error(y_test, model.predict(X_test)))
    print('Test R^2 Score:', 
          r2_score(y_test, model.predict(X_test)))

features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
fit(train, test, features)
```

-with 3 features:

```python
features = ['bedrooms', 'bathrooms', 'longitude']
fit(train, test, features)
```

# What to work on
- [x] Comment more in code 
- [x] Less lines and more organized notebooks 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Nicolae Dubancoe- Regression 1 Sprint Challenge

# Part 1. Predict avocado sales

***
**What went well:**

-Good collective use of imports and reading in data of the :

```python
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)


```
#Part 2. Predict NYC apartment rent

-Nice time series graph:

```python
avocado['Date'] = pd.to_datetime(avocado['Date'])
avocado = avocado.set_index('Date')
avocado.plot(figsize=(16,8));

```

## Importing fbprophet

- Nice work here importing the library and using the .fit method on the avocado_cv dataset:

```python

from fbprophet import Prophet
m = Prophet()
m.fit(avocado)

```


## Prophet Plotting

- Very simple but powerful code! Nice use of comments too :

```python


fig1 = m.plot(forecast)




# Plot components
fig2 = m.plot_components(forecast)



```

- Good use of [cross_validation](https://www.google.com) with Facebook's diagnostics API:

```python

# Create time series cross-validation using the Prophet functionality
# Import needed functions
from fbprophet.diagnostics import cross_validation
avocado_cv = cross_validation(avocado_model, initial='730 days', 
                         horizon='90 days', period='45 days')
```

## Calculating MAPE(Mean Absolute Percentage Error)

- Impressive use of finding the MAPE!:

```python


from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(avocado_cv, metric='mae')


```
# Part 2. Predict NYC apartment rent

## Setting Target

- Nice job setting feature and target of rent:  

```python
# Selecting 2 features to use for linear regression testing
features = ['bedrooms', 'bathrooms']

target = 'price'
```

## MAE

- Excellent work. I mean this because this is important when looking at features:

```python
mean_absolute_error(y_test, y_pred)


```
**What could be better**

- Clean print format like so: 

```python
print('The MAE for test is:', rent_test_mae, '\n')
print('The R^2 for test is:', rent_test_r2)
```

# R^2 Score

- Nice job getting r^2 and y hat out there to work:

```python

r2_score(y_test, y_pred)

```
#What to work on
- [x] Comment more in code
- [x] Come to PM for questions or help
